File: benches/java/src/Harness.java
import java.time.Instant;
import java.util.Arrays;

public final class Harness {
    public static final int HEADER_LEN = 16;

    private Harness() {}

    public static long nowNanos() {
        Instant now = Instant.now();
        return now.getEpochSecond() * 1_000_000_000L + now.getNano();
    }

    public static final class Recorder {
        private final long[] latencies;
        private int count = 0;
        private long highestSeq = -1;
        private long gaps = 0;
        private long reordered = 0;

        public Recorder(int capacity) {
            this.latencies = new long[capacity];
        }

        public void record(long seq, long sentNanos) {
            if (count == latencies.length) {
                return;
            }
            latencies[count++] = Math.max(0, nowNanos() - sentNanos);
            if (highestSeq >= 0) {
                if (seq > highestSeq + 1) {
                    gaps += seq - highestSeq - 1;
                } else if (seq <= highestSeq) {
                    reordered++;
                }
            }
            if (seq > highestSeq) {
                highestSeq = seq;
            }
        }

        public int size() {
            return count;
        }

        private double quantileUs(long[] sorted, double q) {
            if (sorted.length == 0) {
                return 0;
            }
            int index = (int) Math.ceil(q * sorted.length) - 1;
            index = Math.max(0, Math.min(sorted.length - 1, index));
            return sorted[index] / 1000.0;
        }

        public void report(String subject, int payload) {
            if (count == 0) {
                System.out.printf("%s @ %dB: no samples received%n", subject, payload);
                return;
            }
            long[] sorted = Arrays.copyOf(latencies, count);
            Arrays.sort(sorted);
            System.out.printf("subject      %s%n", subject);
            System.out.printf("payload      %d B%n", payload);
            System.out.printf("received     %d%n", count);
            System.out.printf("dropped      %d (gaps in sequence)%n", gaps);
            System.out.printf("reordered    %d%n", reordered);
            System.out.printf("p50          %9.2f us%n", quantileUs(sorted, 0.50));
            System.out.printf("p99          %9.2f us%n", quantileUs(sorted, 0.99));
            System.out.printf("p999         %9.2f us%n", quantileUs(sorted, 0.999));
            System.out.printf("max          %9.2f us%n", sorted[sorted.length - 1] / 1000.0);
        }
    }

    public static final class Pacer {
        private final long intervalNanos;
        private long next;

        public Pacer(long rateHz) {
            this.intervalNanos = 1_000_000_000L / Math.max(1, rateHz);
            this.next = System.nanoTime();
        }

        public void await() {
            next += intervalNanos;
            while (true) {
                long remaining = next - System.nanoTime();
                if (remaining <= 0) {
                    return;
                }
                if (remaining > 1_000_000L) {
                    try {
                        Thread.sleep((remaining - 1_000_000L) / 1_000_000L);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        return;
                    }
                } else {
                    Thread.onSpinWait();
                }
            }
        }
    }
}

File: benches/src/harness.rs
use hdrhistogram::Histogram;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

pub const HEADER_LEN: usize = 16;

pub fn now_nanos() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("system clock is before the unix epoch")
        .as_nanos() as u64
}

pub fn encode(buf: &mut [u8], seq: u64) {
    buf[0..8].copy_from_slice(&seq.to_le_bytes());
    buf[8..16].copy_from_slice(&now_nanos().to_le_bytes());
}

pub fn decode(buf: &[u8]) -> Option<(u64, u64)> {
    if buf.len() < HEADER_LEN {
        return None;
    }
    let seq = u64::from_le_bytes(buf[0..8].try_into().ok()?);
    let sent = u64::from_le_bytes(buf[8..16].try_into().ok()?);
    Some((seq, sent))
}

pub struct Recorder {
    hist: Histogram<u64>,
    received: u64,
    highest_seq: Option<u64>,
    gaps: u64,
    reordered: u64,
}

impl Recorder {
    pub fn new() -> Self {
        Recorder {
            hist: Histogram::new_with_bounds(1, 60_000_000_000, 3)
                .expect("histogram bounds are valid"),
            received: 0,
            highest_seq: None,
            gaps: 0,
            reordered: 0,
        }
    }

    pub fn record(&mut self, seq: u64, sent_nanos: u64) {
        let latency = now_nanos().saturating_sub(sent_nanos);
        self.hist.saturating_record(latency);
        self.received += 1;

        match self.highest_seq {
            None => {}
            Some(highest) if seq > highest + 1 => self.gaps += seq - highest - 1,
            Some(highest) if seq <= highest => self.reordered += 1,
            Some(_) => {}
        }
        if self.highest_seq.is_none_or(|h| seq > h) {
            self.highest_seq = Some(seq);
        }
    }

    pub fn len(&self) -> u64 {
        self.received
    }

    pub fn is_empty(&self) -> bool {
        self.received == 0
    }

    pub fn report(&self, subject: &str, payload: usize) {
        if self.is_empty() {
            println!("{subject} @ {payload}B: no samples received");
            return;
        }
        let us = |v: u64| v as f64 / 1000.0;
        println!("subject      {subject}");
        println!("payload      {payload} B");
        println!("received     {}", self.received);
        println!("dropped      {} (gaps in sequence)", self.gaps);
        println!("reordered    {}", self.reordered);
        println!("p50          {:>9.2} us", us(self.hist.value_at_quantile(0.50)));
        println!("p99          {:>9.2} us", us(self.hist.value_at_quantile(0.99)));
        println!("p999         {:>9.2} us", us(self.hist.value_at_quantile(0.999)));
        println!("max          {:>9.2} us", us(self.hist.max()));
    }
}

impl Default for Recorder {
    fn default() -> Self {
        Self::new()
    }
}

pub struct Pacer {
    interval: Duration,
    next: SystemTime,
}

impl Pacer {
    pub fn new(rate_hz: u64) -> Self {
        Pacer {
            interval: Duration::from_nanos(1_000_000_000 / rate_hz.max(1)),
            next: SystemTime::now(),
        }
    }

    pub fn wait(&mut self) {
        self.next += self.interval;
        loop {
            let Ok(remaining) = self.next.duration_since(SystemTime::now()) else {
                return; // deadline already passed
            };
            if remaining > Duration::from_millis(1) {
                std::thread::sleep(remaining - Duration::from_millis(1));
            } else {
                std::hint::spin_loop();
            }
        }
    }
}
